File: projects/m1/046-time-calculator/solutions/andreaattardo/timecalculator.cpp
/*
Time Calculator
Write a function named add_time that takes in two required parameters and one optional parameter:
-a start time in the 12-hour clock format (ending in AM or PM)
-a duration time that indicates the number of hours and minutes
-(optional) a starting day of the week, case insensitive

The function should add the duration time to the start time and return the result.
If the result will be the next day, it should show (next day) after the time. If the result will be more than one day 
later, it should show (n days later) after the time, where "n" is the number of days later.
If the function is given the optional starting day of the week parameter, then the output should display the day of the 
week of the result. The day of the week in the output should appear after the time and before the number of days later.
Below are some examples of different cases the function should handle. Pay close attention to the spacing and punctuation
of the results.

add_time("3:00 PM", "3:10")
# Returns: 6:10 PM

add_time("11:30 AM", "2:32", "Monday")
# Returns: 2:02 PM, Monday

add_time("11:43 AM", "00:20")
# Returns: 12:03 PM

add_time("10:10 PM", "3:30")
# Returns: 1:40 AM (next day)

add_time("11:43 PM", "24:20", "tueSday")
# Returns: 12:03 AM, Thursday (2 days later)

add_time("6:30 PM", "205:12")
# Returns: 7:42 AM (9 days later)
*/


#include <iostream>
#include <string>
#include "timecalculator.h"

using namespace std;


//MAIN
int main(){

	cout << "TIME CALCULATOR." << endl;

	//init times value from usr
	string startTime, durationTime, startingDay, endTime, e;
	getTimes(startTime, durationTime, startingDay);
	
	//time conversion in minutes
	int startTMin = timeToMin(startTime); 
	int durTMin = timeToMin(durationTime); 
	int endTMin = startTMin + durTMin;
	//conversion starting day name in number
	int startDayNum = dayNameToNum(startingDay);

	//check if ending day has surpassed one or more day (1440 min = 24h)
	while ( (endTMin - 1440) >= 0 ) {
		endTMin -= 1440;
		startDayNum++;
	}
	//e.g. write 12:05 instead of 00:05
	if (endTMin < 60) endTMin += 720;

	startingDay = numToDayName(startDayNum);

	endTime = minToTime(endTMin);
	cout << endTime << " " << startingDay << endl;

	return 0;
}



//FUNCTIONS DEFINITIONS

void getTimes(string& startTime, string& durationTime, string& startingDay) {

	cout << "Insert a start time in 12 hour format (e.g. 7:00 PM): " << endl;
	getline(cin, startTime);

	cout << "Insert a duration time in hours:minutes format (e.g. 16:45): " << endl;
	getline(cin, durationTime);

	cout << "Do you like to insert a starting day (y/n)?" << endl;
	cin >> startingDay;
	if (startingDay == "y")
	{
		do {
			cout << "Insert starting day (e.g. Monday):" << endl;
			cin >> startingDay;
		} while (!isLegalDayName(startingDay));

	}
	else startingDay = " ";
}


int timeToMin(string& time) {
	int timeInMinutes = 0;
	time.erase(time.find(':'), 1);
	timeInMinutes = stoi(time);
	timeInMinutes = (timeInMinutes / 100 * 60) + (timeInMinutes % 100);

	if (time.find('P') != string::npos && timeInMinutes <= 720)
	{
		timeInMinutes += 720;
	}

	return timeInMinutes;
}


string minToTime(int& timeInMinutes) {
	string time;
	string xM = "AM";

	if (timeInMinutes >= 720) xM = "PM";
	if (timeInMinutes >= 780) timeInMinutes -= 720;

	time = to_string(timeInMinutes / 60) + ":";
	if ((timeInMinutes % 60) < 10) time += "0";
	time += to_string(timeInMinutes % 60) + " " + xM;

	return time;
}


string numToDayName(int& dayNum) {
	if ( dayNum < 0 ) return " ";
	//compute if dayNum is more then one week in which day is
	if (dayNum > 6) dayNum %= 7;

	string weekday[7] = { "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday" };

	return weekday[dayNum];
}


int dayNameToNum(string& dayName) {
	
	string weekday[7] = { "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday" };

	for (int i = 0; i < size(weekday); i++) {
		if (weekday[i] == dayName) return i;
	}

	return -1;
}

bool isLegalDayName(string& dayName) {
	string weekday[7] = { "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday", "Sunday" };

	for (int i = 0; i < size(weekday); i++) {
		if (weekday[i] == dayName) return true;
	}

	return false;
}

File: projects/m1/046-time-calculator/solutions/RoccoFigliamonti/time_calculator.py
"""
Write a function named `add_time` that takes in two required parameters and one optional parameter:
* a start time in the 12-hour clock format (ending in AM or PM) 
* a duration time that indicates the number of hours and minutes
* (optional) a starting day of the week, case insensitive

The function should add the duration time to the start time and return the result.

If the result will be the next day, it should show `(next day)` after the time. If the result will be more than one day later, it should show `(n days later)` after the time, where "n" is the number of days later.

If the function is given the optional starting day of the week parameter, then the output should display the day of the week of the result. The day of the week in the output should appear after the time and before the number of days later."""


def add_time(start_time, duration_time, starting_day =""):

    #parsing
    start_hour = int(start_time.split(":")[0])
    start_minutes = int((start_time.split(":")[1].split()[0]))
    start_am_or_pm = (start_time.split(":")[1].split()[1])
    duration_hour = int(duration_time.split(":")[0])
    duration_minutes = int(duration_time.split(":")[1])
    final_am_or_pm = ""

    #new time
    new_minutes = (start_minutes + duration_minutes)%60 #resto della somma dei minuti
    new_hours = ((start_hour+duration_hour) + (start_minutes+duration_minutes)//60)%12 #resto della somma delle ore + del quozionte della somma dei minuti

    #AM or PM 
    if (start_am_or_pm == "AM" and \
       (((start_hour+duration_hour) + (start_minutes+duration_minutes)//60)//12)%2 == 0) \
        or \
        (start_am_or_pm == "PM" and \
       (((start_hour+duration_hour) + (start_minutes+duration_minutes)//60)//12)%2 != 0):
        final_am_or_pm = "AM"
    else:
        final_am_or_pm = "PM"
    
    #counting days passed
    if start_am_or_pm == "AM":
        days_gone = ((start_hour+duration_hour) + (start_minutes+duration_minutes)//60)//24
    else:
        days_gone = ((12+start_hour+duration_hour) + (start_minutes+duration_minutes)//60)//24

    #days optional parameter
    calendar = {"monday":0, "tuesday":1, "wednesday":2, "thursday":3, "friday":4, "saturday":5, "sunday":6}
    if starting_day.lower() in calendar.keys():
        new_day_value = (calendar[starting_day.lower()] + days_gone)%7
        new_calendar = {v:k for k,v in calendar.items()}
        new_day_key = new_calendar[new_day_value]

    #final output format
    if days_gone == 0 and starting_day !="":
        time_mex = (f"{new_hours}:{new_minutes:02d} {final_am_or_pm}, {new_day_key}")
    elif days_gone == 0:
        time_mex = (f"{new_hours}:{new_minutes:02d} {final_am_or_pm} ")
    elif days_gone == 1 and starting_day !="":
        time_mex = (f"{new_hours}:{new_minutes:02d} {final_am_or_pm}, {new_day_key} (next day)")
    elif days_gone == 1:
        time_mex = (f"{new_hours}:{new_minutes:02d} {final_am_or_pm} (next day)")
    elif days_gone > 1 and starting_day !="":
        time_mex = (f"{new_hours}:{new_minutes:02d} {final_am_or_pm}, {new_day_key} ({days_gone} days later)")
    else:
        time_mex = (f"{new_hours}:{new_minutes:02d} {final_am_or_pm} ({days_gone} days later)")
    
    return time_mex


print(add_time("3:00 PM", "3:10"))
# Returns: 6:10 PM

print(add_time("11:30 AM", "2:32", "Monday"))
# Returns: 2:02 PM, Monday

print(add_time("11:43 AM", "00:20"))
# Returns: 12:03 PM

print(add_time("10:10 PM", "3:30"))
# Returns: 1:40 AM (next day)

print(add_time("11:43 PM", "24:20", "tueSday"))
# Returns: 12:03 AM, Thursday (2 days later)

print(add_time("6:30 PM", "205:12"))
# Returns: 7:42 AM (9 days later)

